examples/chatbot/src/websocket.rs

```diff
@@ -0,0 +1,175 @@
+                    )) => {
+                        tracing::warn!(
+                            "connection was sent an unexpected frame or was reset, reestablishing it"
+                        );
+                        s = self
+                            .connect().instrument(span)
+                            .await
+                            .context("when reestablishing connection")?;
+                        continue
+                    }
+                    _ => msg.context("when getting message")?,
+                };
+                self.process_message(msg, &mut event_fn).instrument(span).await?
+            })
+        }
+    }
+
+    /// Process a message from the websocket
+    pub async fn process_message<Fut>(
+        &mut self,
+        msg: tungstenite::Message,
+        event_fn: &mut impl FnMut(Event, types::Timestamp) -> Fut,
+    ) -> Result<(), eyre::Report>
+    where
+        Fut: std::future::Future<Output = ()>,
+    {
+        match msg {
+            tungstenite::Message::Text(s) => {
+                tracing::trace!("{s}");
+                // Parse the message into a [twitch_api::eventsub::EventsubWebsocketData]
+                match Event::parse_websocket(&s)? {
+                    EventsubWebsocketData::Welcome {
+                        payload: WelcomePayload { session },
+                        ..
+                    }
+                    | EventsubWebsocketData::Reconnect {
+                        payload: ReconnectPayload { session },
+                        ..
+                    } => {
+                        self.process_welcome_message(session).await?;
+                        Ok(())
+                    }
+                    EventsubWebsocketData::Notification { metadata, payload } => {
+                        event_fn(payload, metadata.message_timestamp.into_owned()).await;
+                        Ok(())
+                    }
+                    EventsubWebsocketData::Revocation {
+                        metadata,
+                        payload: _,
+                    } => eyre::bail!("got revocation event: {metadata:?}"),
+                    EventsubWebsocketData::Keepalive {
+                        metadata: _,
+                        payload: _,
+                    } => Ok(()),
+                    _ => Ok(()),
+                }
+            }
+            tungstenite::Message::Close(_) => todo!(),
+            _ => Ok(()),
+        }
+    }
+
+    pub async fn process_welcome_message(
+        &mut self,
+        data: SessionData<'_>,
+    ) -> Result<(), eyre::Report> {
+        tracing::info!("connected to twitch chat");
+        self.session_id = Some(data.id.to_string());
+        if let Some(url) = data.reconnect_url {
+            self.connect_url = url.parse()?;
+        }
+        let mut token = self.token.lock().await;
+        // check if the token is expired, if it is, request a new token. This only works if using a oauth service for getting a token
```

Comment:
> This only works if using a oauth service for getting a token

I'm not sure, but this says "We can refresh, because we're using device code grant", right?

Reply:
right, copy pasted from the other example for websocket and forgot to change :)